docs/manual/src/internals/foreign_calls.md

```diff
@@ -0,0 +1,88 @@
+# Rust -> Foreign calls
+
+How can Rust code make a call into the foreign language?
+The [lifting and lowering](./lifting_and_lowering.md) docs describe this at a high-level, this document will explore the details.
+
+This document only describes non-async calls.
+See the [Async FFI](./async-ffi.md) docs for a description of async calls.
+
+## VTables
+
+Calling into foreign code has 2 major differences from [calling Rust code](./rust_calls.md):
+
+* Foreign code is only called via an interface.
+* The Rust code does not have access to a library with functions it can call.
+
+Because of this, the foreign calls all go through a vtable that the foreign code registers with the Rust code.
+This is a `repr(C)` struct where each entry is a function pointer corresponding to a method in a callback interface.
+
+For example:
+
+```rust
+/// For this trait interface...
+#[uniffi::export(with_foreign)]
+pub trait TodoList {
+    fn append(&self, title: String);
+    fn get_items(&self) -> Vec<String>;
+}
+
+/// A VTable like this will be generated:
+#[repr(C)]
+pub struct UniffiTodoListVTable {
+    // One function pointer field for each method
+    add: extern "C" fn(uniffi_handle: u64, title: RustBuffer, uniffi_out_return: &mut (), uniffi_call_status: &mut RustCallStatus),
+    get_items: extern "C" fn(uniffi_handle: u64, uniffi_out_return: &mut RustBuffer, uniffi_call_status: &mut RustCallStatus);
+    // A function pointer for freeing the callback interface
+    uniffi_free: extern "C" fn(uniffi_handle: u64);
+}
+
+/// A function to register the VTable
+#[no_mangle]
+pub extern "C" fn uniffi_init_todo_list_vtable(vtable: std::ptr::NonNull<UniffiTodoListVTable>) {...}
+```
+
+
+## VTable methods
+
+VTable methods work similar to a [Rust method call](./rust_calls.md)], with some exceptions.
+
+Instead of returning the value as normal, it's written to an out pointer.
+This is because some foreign languages, like Python, have issues returning C structs like `RustBuffer`.
+Methods that don't return anything still have an out pointer argument, but it's never written to.
+
+The first argument is a `u64` handle rather than a `void *`.
```

Comment:
Sorry, but I also don't understand why it would be `void *` in the first place :)

Reply:
I was trying to contrast and compare callback interfaces with "Rust interfaces" (i.e. Rust structs where we expose their methods).  I don't think it's that useful though, especially since I'm hoping to make them use the same handle type, so I just removed the section.